osquery/tables/system/darwin/xattr.cpp

```diff
@@ -0,0 +1,166 @@
+  if(field.header_length > 8){field.header_length = 0; return field;}
+  
+  for(unsigned int i = 1; i < field.header_length+1; i++){
+    field.length = field.length << 8;
+    field.length += (unsigned char)parse[buffer_position+i];
+  }
+  return field;
+}
+
+std::string fixString(const std::string& toFix){
+  std::stringstream result;
+  unsigned char byte;
+  int count = 0;
+  for (int i = 0; i < toFix.length(); ++i) {
+    byte = toFix[i];
+    if ((int)byte > 0x1F && (int)byte < 0x80) {
+      result << byte;
+      continue;
+    } else if (byte == 0) {
+      result << ' ';
+    } else {
+      result << '%' << std::setfill('0') << std::setw(2) << std::hex << (int)byte;
+    }
+    count++;
+  }
+  VLOG(1) << "Had " << count << " control characters to remove"; 
+  return result.str();
+}
+
+void parseWhereFromData(Row &r, const std::string& parse, int ret, int buf_len){
+  if(ret == -1){
+    VLOG(1) << handleError();
+  }else{
+    r["raw64"] = base64Encode(parse);
+    if(buf_len < 11 || 0x5F != (unsigned char)parse[11]){
+      r["download_url"] = "No data";
+      r["download_page"] = "No data";
+    }else{
+      unsigned int starting_position = 12;
+      struct XAttrField field = getFieldLength(starting_position, parse, buf_len);
+      starting_position += 1+field.header_length;
+      if(starting_position + field.length >= parse.length()){return;}
+      r["download_url"] = fixString(parse.substr(starting_position,field.length));
+      starting_position += field.length+1;
+      if(starting_position + field.length >= parse.length()){return;}
+      field = getFieldLength(starting_position, parse, buf_len);
+      starting_position += field.header_length+1;
+      r["download_page"] = fixString(parse.substr(starting_position, field.length));
+    }
+  }
+}
+
+void getFileData(Row& r, const boost::filesystem::path path){
+  r["path"] = path.string();
+  r["directory"] = path.parent_path().string();
+  int ret;
+  int buf_len;
+  std::string att_str = getAttribute(path, "com.apple.metadata:kMDItemWhereFroms", ret, buf_len);
+  parseWhereFromData(r, att_str, ret, buf_len);
+}
+
+QueryData genXattr(QueryContext& context) {
+  QueryData results;
+  auto paths = context.constraints["path"].getAll(EQUALS);
+
+  for (const auto& path_string : paths) {
+    boost::filesystem::path path = path_string;
+    if (!boost::filesystem::is_regular_file(path)) {
+      continue;
+    }
+    Row r;
+    getFileData(r, path);
+    results.push_back(r);
+  }
+
+  auto directories = context.constraints["directory"].getAll(EQUALS);
+  for (const auto& directory_string : directories) {
+    boost::filesystem::path directory = directory_string;
+    if (!boost::filesystem::is_directory(directory)) {
+      continue;
+    }
+
+    // Iterate over the directory and generate a hash for each regular file.
+    boost::filesystem::directory_iterator begin(directory), end;
```

Comment:
I think there's a helper function in `<osquery/filesystem.h>` for listing directories. It takes care of all the iterator exceptions you're not handling too. ;) SIGABRT!!!